src/v/cloud_storage/partition_manifest.cc

```diff
@@ -305,85 +309,420 @@ partition_manifest::difference(const partition_manifest& remote_set) const {
+        expect_segment_meta_start,
+        expect_segment_meta_key,
+        expect_segment_meta_value,
+        terminal_state,
+    } _state{state::expect_manifest_start};
+
+    using segment_map = partition_manifest::segment_map;
+
+    key_string _manifest_key;
+    key_string _segment_meta_key;
+    segment_name _segment_name;
+    std::optional<segment_name_components> _parsed_segment_key;
+    partition_manifest::key _segment_key;
+    partition_manifest::segment_meta _meta;
+    std::unique_ptr<segment_map> _segments;
+
+    // required manifest fields
+    std::optional<int32_t> _version;
+    std::optional<model::ns> _namespace;
+    std::optional<model::topic> _topic;
+    std::optional<int32_t> _partition_id;
+    std::optional<model::initial_revision_id> _revision_id;
+    std::optional<model::offset> _last_offset;
+
+    // required segment meta fields
+    std::optional<bool> _is_compacted;
+    std::optional<size_t> _size_bytes;
+    std::optional<model::offset> _base_offset;
+    std::optional<model::offset> _committed_offset;
+
+    // optional segment meta fields
+    std::optional<model::timestamp> _base_timestamp;
+    std::optional<model::timestamp> _max_timestamp;
+    std::optional<model::offset> _delta_offset;
+    std::optional<model::initial_revision_id> _ntp_revision;
+    std::optional<model::term_id> _archiver_term;
+
+    void check_that_required_meta_fields_are_present() {
+        if (!_is_compacted) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Missing _is_compacted value in {} segment meta",
+              _segment_name));
+        }
+        if (!_size_bytes) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Missing _size_bytes value in {} segment meta",
+              _segment_name));
+        }
+        if (!_base_offset) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Missing _base_offset value in {} segment meta",
+              _segment_name));
+        }
+        if (!_committed_offset) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "Missing _committed_offset value in {} segment meta",
+              _segment_name));
+        }
+    }
+
+    void clear_meta_fields() {
+        // required fields
+        _is_compacted = std::nullopt;
+        _size_bytes = std::nullopt;
+        _base_offset = std::nullopt;
+        _committed_offset = std::nullopt;
+
+        // optional segment meta fields
+        _base_timestamp = std::nullopt;
+        _max_timestamp = std::nullopt;
+        _delta_offset = std::nullopt;
+        _ntp_revision = std::nullopt;
+        _archiver_term = std::nullopt;
+    }
+
+    void check_manifest_fields_are_present() {
+        if (!_version) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format, "Missing _version value partition manifest"));
```

Comment:
nit: in practice, if this indicates a missing `"version"` field in the manifest, maybe use `"version"` instead of `_version`?

Reply:
fixed